src/mongo/delegates/Updates.class.php

```diff
@@ -478,61 +479,87 @@ protected function applyChangeSet(\Tripod\ChangeSet $cs, $originalCBDs, $context
                     {
                         $subjectsAndPredicatesOfChange[$subjectOfChange][] = $predicate;
                     }
 
-                    $object = $cs->get_subject_property_values($r["value"],$this->labeller->qname_to_uri("rdf:object"));
-
-                    $isUri = ($object[0]['type']=="uri");
-
-                    // is $predicate in $docs already?
-                    $valueExists = ($isUri) ? $targetGraph->has_resource_triple($subjectOfChange,$predicate,$object[0]["value"]) : $targetGraph->has_literal_triple($subjectOfChange,$predicate,$object[0]["value"]);
-
-                    if (!$valueExists)
+                    // set to existing object if exists
+                    $valueObject = ($predicateExists) ? $doc[$nsPredicate] : array();
+                    if (isset($valueObject[VALUE_URI]) || isset($valueObject[VALUE_LITERAL]))
                     {
-                        $this->errorLog("Removal value {$subjectOfChange} {$predicate} {$object[0]['value']} does not appear in target document to be updated",array("targetGraph"=>$targetGraph->to_ntriples()));
-                        throw new \Exception("Removal value {$subjectOfChange} {$predicate} {$object[0]['value']} does not appear in target document to be updated");
+                        // this is a single value object, convert to array of values for now
+                        $valueObject = array($valueObject);
                     }
-                    else if ($isUri)
-                    {
-                        $targetGraph->remove_resource_triple($subjectOfChange,$predicate,$object[0]["value"]);
 
-                    }
-                    else
+                    if (isset($additionsRemovals["additions"]))
                     {
-                        $targetGraph->remove_literal_triple($subjectOfChange,$predicate,$object[0]["value"]);
-
+                        foreach ($additionsRemovals["additions"] as $addition)
+                        {
+                            $valueObject[] = $addition;
+                        }
                     }
 
-                }
-
-                foreach ($additions as $r)
-                {
-                    $predicate = $cs->get_first_resource($r["value"],$this->labeller->qname_to_uri("rdf:predicate"));
-
-                    if(!in_array($predicate, $subjectsAndPredicatesOfChange[$subjectOfChange]))
+                    // remove
+                    if (isset($additionsRemovals["removals"]))
                     {
-                        $subjectsAndPredicatesOfChange[$subjectOfChange][] = $predicate;
+                        $elemsToRemove = array();
+                        foreach ($additionsRemovals["removals"] as $removal)
+                        {
+                            foreach($removal as $valueType => $v) { // should only have one k/v
+                                $found = false;
+                                for ($i=0; $i<count($valueObject);$i++) {
```

Comment:
I think all of this could be simplified to:
```php
foreach($removal as $r)
{
    $valueIndex = array_search($r, $valueObject);
    if($valueIndex === false) 
    {
        $v = array_pop(array_values($r));
        $this->errorLog("Removal value {$subjectOfChange} {$predicate} {$v} does not appear in target document to be updated",array("doc"=>$doc));
        throw new \Exception("Removal value {$subjectOfChange} {$predicate} {$v} does not appear in target document to be updated");
    }
    else
    {
        $elemsToRemove[] = $valueIndex;
    }
}